### PR TITLE
Length modes

### LINKED ISSUE #40: Feature request: UTF-16 code units length validator
(Adjusted for `garde` from https://github.com/Keats/validator/issues/250)

The current `garde` crate provides built-in validators for various use cases, but it lacks a validator for checking the length of a string based on its UTF-16 code units. This feature request proposes the addition of a UTF-16 code units length validator to the crate.

The motivation behind this request stems from the need to match the behavior of the HTML textarea `maxlength` attribute, which counts UTF-16 code units. To provide better consistency between frontend and backend validation, it would be useful to have a validator that directly checks the length of a string based on its UTF-16 code units.

The new validator could be used as follows:

```rust
use garde::{Validate, Valid};

#[derive(Validate)]
struct Post<'a> {
    #[garde(length_utf16(min=32, max=65535))]
    message: &'a str,,
}
```

Replace `N` with the desired UTF-16 code unit count. Use the same `N` for the HTML `textarea`. 

`N` would be `max bytes / 2`, as UTF-16 code units are 2 bytes long.

This new validator would ensure that the code unit count limits are consistent between the HTML textarea and Rust, despite the different character encodings used, and avoid false negatives.

### LINKED ISSUE #84: Rework string length rules
Hello! Thank you for your crate. I highly appreciate it. However, I've noticed one aspect throughout the entire crate that doesn't resonate with me - specifically, the implementation of the string `length` rule. In my view, this rule could be misleading because:

> It's important to remember that char represents a Unicode Scalar Value, and may not match your idea of what a 'character' is. Iteration over grapheme clusters may be what you actually want.

(Referenced from [here](https://stackoverflow.com/questions/46290655/does-rusts-string-have-a-method-that-returns-the-number-of-characters-rather-th))

One reason I admire Rust is its handling of strings: by default, it provides size in bytes, and additionally, you can get the size in **Unicode Scalar Values** using the `.chars().count()` methods, which doesn't always(!) *match your idea of what a "character"* (as mentioned above). It seemed unfamiliar to me that the `length` rule doesn't return the size in bytes.

Therefore, I would like to suggest a new approach for strings:
- Move the `length` (`.chars().count()`) implementation to the `char_count` rule
- Create a new `grapheme_count` rule that will available with own feature flag because it will require a crate for working with graphemes (e.g. [unicode-segmentation](https://github.com/unicode-rs/unicode-segmentation)).
- (Optional) Move the `byte_length` implementation to the `length` rule

Downsides of the optional solution: it will significantly break backward compatibility.

### PR DESCRIPTION
**Pull request by jprochazk**
- Closes https://github.com/jprochazk/garde/issues/84
- Closes https://github.com/jprochazk/garde/issues/40

- Add a mode argument to the `length` rule. Currently supports:
  - `simple`, which is the default
  - `bytes`
  - `chars`
  - `graphemes`
  - `utf16`
- Remove `byte_length`

### CODE REVIEW

**Comment by jprochazk** [comment]
`half` is causing way too many issues. Don't want to bump the msrv, so I'm just going to remove the benchmarks and `criterion`.

File: garde/benches/validation.rs
Comment (aumetra) [inline garde/benches/validation.rs:1]: benchmarks are over 😔

**Review by aumetra — APPROVED** [review]
Seems fine to me

**Comment by jprochazk** [comment]
Released in [v0.18](https://github.com/jprochazk/garde/releases/tag/v0.18.0)

**Pull request merged**